### DISCUSSION TITLE
Google Season of Docs 2024

### DISCUSSION THREAD

**Discussion by AlbumenJ** [apache/dubbo#14010] [Notice]
# Update Apache Dubbo Triple Protocol Documentations

## About us
Apache Dubbo is a robust RPC service development framework that has been thoughtfully designed to effectively address the myriad service governance and communication challenges that are often encountered in a microservice architecture environment. It proudly offers official support for multiple language SDKs, with Java and Golang included among them.

When you develop microservices with the help of Dubbo, you'll benefit from native support for tasks such as remote address discovery and intercommunication between various services. This becomes possible because of Dubbo's rich and comprehensive set of service governance features. These features are specially designed to help you meet a wide range of requirements, including, but not limited to, service discovery, load balancing, and the scheduling of traffic.

One of the standout aspects of Dubbo is its high level of scalability. This is an attribute that empowers users to easily implement custom logic for tasks such as traffic interception and the selection of locations. This makes Dubbo an ideal solution for organizations that are seeking a flexible and adaptable framework for their microservice architecture.

## Problem

The HTTP protocol plays a pivotal role in the Remote Procedure Call (RPC) framework. It is essential in facilitating communication between programs across different network systems. Dubbo, a high-performance, Java-based RPC framework, has recognized the significance of the HTTP protocol.

Starting from version 3.0.x, Dubbo has implemented a new protocol, the Triple protocol. It has been designed to be fully compatible with the gRPC protocol. This compatibility has had profound implications as it enhances the interoperability between different components of the system, thereby allowing for more efficient and reliable communication between various services.

The Triple protocol's compatibility with gRPC means that it not only enhances the communication within Dubbo, but also allows Dubbo to communicate effectively with other systems using the gRPC protocol. This cross-system communication is a boon for systems that are designed around microservices, as it allows for seamless communication and data exchange between different services, regardless of the system they belong to.

The introduction of the Triple protocol has also enabled Dubbo to provide the feature of bi-directional streaming. This is a significant feature as it allows for real-time communication between client and server, which can greatly enhance the responsiveness and efficiency of the system. The bi-directional streaming feature, combined with the enhanced interoperability provided by the Triple protocol, has made Dubbo a highly efficient and reliable system for managing microservices.

Since the introduction of the Triple protocol, we have received many positive responses about Dubbo. Users have praised its enhanced capabilities, interoperability, and efficient communication, highlighting the significant impact of the Triple protocol on the performance and reliability of Dubbo.

### Task 1 : Troubleshooting documents

While we have provided the triple itself, we find that we are lacking in troubleshooting documents. This is a significant issue because it means that our users are left to grapple with the complexities and nuances of the HTTP/2 protocol on their own. Many of these users are experiencing difficulties, such as the connection closing for reasons that they cannot understand or troubleshoot. This is a clear indicator that we need to focus our efforts on developing robust and comprehensive troubleshooting documents. With these in place, we can better support our users and empower them to navigate the challenges associated with the HTTP/2 protocol.

### Task 2 : Design and usage documents for brand new feature

We've gone to great lengths to provide the Triple protocol, an advanced protocol similar to gRPC and based on HTTP/2. However, despite its advancements, it's important to acknowledge that it still functions as a RAW data-like protocol, making it less user-friendly and more challenging to navigate. This complexity presents difficulties in debugging for users and may even render it inaccessible for frontend developers to call in a browser.

Recognizing these challenges, we've identified the need to improve and evolve the Triple protocol. Our goal is to enhance its capabilities and make it as intuitive and easy to use as RESTful API. To achieve this, we've taken strides and have begun implementing this improvement in the beta version of 3.3.x. The initial results have been promising, demonstrating the feasibility and potential of this enhancement.

However, we understand that providing a new feature is not enough. It's equally crucial to enable our users to understand and utilize this feature fully. Therefore, our next step involves creating comprehensive documentation. This documentation aims to guide users through the process of using this newly enhanced Triple protocol, ensuring they can leverage all the benefits of this amazing feature.

### Task 1 : Troubleshooting documents

- Begin by visiting the FAQ page on the official Apache Dubbo website at [FAQ link](https://dubbo.apache.org/en/docs3-v2/java-sdk/faq/) Thoroughly review all the information presented and evaluate it in terms of readability and detailed explanation. Ensure that the information provided is correct and up-to-date. For further clarification and understanding, all the error codes can be found and cross-checked at [apache/dubbo](https://github.com/apache/dubbo).
- In order to better assist users and improve their experience, it is suggested to introduce some new comprehensive documents. These documents should include detailed guidance on how to effectively use Triple, a key feature of Apache Dubbo. Additionally, provide a step-by-step troubleshooting guide that users can follow should they encounter any common errors while using Triple. This additional resource will equip users with the necessary knowledge to resolve issues independently and efficiently.

### Task 2 : Design and usage documents for brand new feature

- Thoroughly review and revise the design document to incorporate this new feature. Ensure that the description of the feature is clear, straightforward, and concise. This will make it easier for users to comprehend the functionality of the feature and understand the benefits it brings to their user experience.
- After revision, publish comprehensive and detailed documents to guide users on how to effectively utilize this new feature. It's crucial to emphasize that this feature is highly flexible, offering a wide variety of usage possibilities that can be tailored to the specific needs of each user.
- In addition to this, provide a comprehensive troubleshooting guide related to the new feature. This guide should cover potential issues that users may encounter while using the feature and provide effective solutions. This will ensure that users can quickly resolve any problems and get the most out of the new functionality.

## Success Metrics

Here are the success metrics for our projects:

- Increase the percentage of users who can resolve issues by consulting the FAQ only once
- Ensure users can understand the Triple protocol clearly by reading just one document
- Prevent a significant increase in user inquiries after the release of a new feature. All potential questions should be answered in the documentation.

## Preferred skills

Past experience with Java and HTTP

## Timeline

Dates | Action Items
-- | --
May | Orientation
June - August | Audit existing documentation and learning source code
September - October | Create documentation
November | Project completion

## Project budget
Budget item | Amount | Running Total | Notes/justifications
-- | -- | -- | --
Technical writer audit, update, test, and publish new documentation | 4000 | 8000 | 2 writer x 4000 each
Volunteer stipends | 500 | 9000 | 2 volunteer stipends x 500 each
Project SWAG including T-shirts, stickers and shipping fee	 | 150 | 9150 |  
TOTAL |   | 9150 |

**Comment by yashpandey06** [comment]
great !!

**Comment by Ayush9026** [comment]
Awesome Projects.

**Comment by civplayer** [comment]
I am interested, I am currently enrolled in a Masters of Science of Techincal Writing at my university. How do I apply?
